### ISSUE TITLE
Week 10

### ISSUE THREAD

**Issue by Alberto544-cell** [mario-gellrich-zhaw/data_analytics#26]
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>k-means_clustering_task01</title>
  <style>
    body {
      font-family: system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif;
      margin: 20px;
      background: #f6f7fb;
      color: #222;
      line-height: 1.6;
    }
    h1, h2, h3 {
      color: #1b3a57;
    }
    .cell {
      background: #ffffff;
      border-radius: 6px;
      padding: 15px 18px;
      margin-bottom: 16px;
      box-shadow: 0 2px 4px rgba(15, 23, 42, 0.08);
    }
    .cell-title {
      font-size: 0.8rem;
      text-transform: uppercase;
      letter-spacing: 0.06em;
      color: #667085;
      margin-bottom: 6px;
    }
    .markdown {
      border-left: 4px solid #2563eb33;
    }
    .code {
      border-left: 4px solid #16a34a33;
    }
    pre {
      background: #0f172a;
      color: #e5e7eb;
      padding: 12px 14px;
      border-radius: 4px;
      overflow-x: auto;
      font-size: 0.9rem;
    }
    code {
      font-family: "Fira Code", ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
    }
    .note {
      background: #e0f2fe;
      border-left: 4px solid #0ea5e9;
      padding: 10px 12px;
      border-radius: 4px;
      font-size: 0.9rem;
      margin-top: 6px;
    }
  </style>
</head>
<body>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h1>Week 10 – Task 1: K-means clustering</h1>
  <p>
    This HTML document contains my work for <strong>Task 1</strong> of Week 10
    (k-means clustering, elbow method, silhouette analysis, image segmentation,
    and 3D point cloud segmentation).
  </p>
  <p>
    The code cells below are intended to be used inside the original
    <code>k-means_clustering.ipynb</code> notebook provided in the course.
    I added the required changes and explanations for sub-tasks (c)–(h).
  </p>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>1.c) Silhouette analysis for different values of k</h2>
  <p>
    In this part, I compute the <strong>Silhouette Score</strong> for values of
    <em>k</em> ranging from 2 to 11. The dataset <code>X</code> is the same one
    used in the simple k-means example in the original notebook.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

silhouette_scores = []

for i in range(2, 12, 1):
    kmeans = KMeans(n_clusters=i, random_state=42)
    kmeans.fit(X)
    labels = kmeans.labels_
    score = silhouette_score(X, labels)
    silhouette_scores.append(score)
    print(f"k={i}, Silhouette Score = {score}")</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p><strong>Interpretation:</strong></p>
  <ul>
    <li>I look for the value of <em>k</em> with the highest Silhouette Score.</li>
    <li>
      In my run, the best score is around <strong>k = 3–4</strong>, which is
      consistent with the result obtained from the elbow method in the simple
      k-means example.
    </li>
  </ul>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>1.d) Image segmentation with my own JPG image</h2>
  <p>
    Here I use my own image file named <code>my_image.jpg</code>. The image is
    reshaped to a 2D array of RGB values and clustered with k-means. Then I
    reconstruct a segmented image from the cluster centers.
  </p>
  <div class="note">
    <strong>Note:</strong> the file <code>my_image.jpg</code> must be placed in
    the same directory as the notebook before running this code.
  </div>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>import imageio
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans

# Read image (make sure 'my_image.jpg' is in the same folder as this notebook)
img = imageio.imread("my_image.jpg")

# Reshape image to (num_pixels, 3)
img_reshape = img.reshape((-1, 3))

# Example with k = 4 clusters
kmeans_img = KMeans(n_clusters=4, random_state=0).fit(img_reshape)
segmented = kmeans_img.cluster_centers_[kmeans_img.labels_]
segmented = segmented.reshape(img.shape).astype(int)

plt.figure(figsize=(5, 5))
plt.imshow(segmented)
plt.title("Segmented image (k = 4)")
plt.axis("off")
plt.show()</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>1.e) Elbow method for image segmentation</h2>
  <p>
    Next, I use the elbow method on <code>img_reshape</code>, the reshaped
    image data, to estimate the optimal number of clusters for image
    segmentation.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>inertia_list = []
K = range(2, 12)

for k in K:
    kmeans = KMeans(n_clusters=k, random_state=0)
    kmeans.fit(img_reshape)
    inertia_list.append(kmeans.inertia_)

plt.figure()
plt.plot(K, inertia_list, marker="o")
plt.xlabel("k")
plt.ylabel("Inertia")
plt.title("Elbow Method - Image Segmentation")
plt.show()</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p><strong>Interpretation:</strong></p>
  <ul>
    <li>
      I visually inspect the elbow plot and look for the point where the
      decrease in inertia becomes less steep.
    </li>
    <li>
      This point indicates the optimal number of clusters. For natural images,
      this is often somewhere around <strong>4–6 clusters</strong>.
    </li>
  </ul>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>1.f) Four different segmentations of the same image</h2>
  <p>
    In this part, I generate four different image segmentations of the same
    image using different values of <em>k</em>.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>for k in [2, 4, 6, 8]:
    kmeans = KMeans(n_clusters=k, random_state=0).fit(img_reshape)
    seg = kmeans.cluster_centers_[kmeans.labels_]
    seg = seg.reshape(img.shape).astype(int)

    plt.figure(figsize=(5, 5))
    plt.imshow(seg)
    plt.title(f"Segmented image (k = {k})")
    plt.axis("off")
    plt.show()</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p><strong>Observation:</strong></p>
  <ul>
    <li>
      With small values of <em>k</em> (e.g. 2), the segmentation is very coarse
      and only a few large color regions are visible.
    </li>
    <li>
      For higher values of <em>k</em> (e.g. 8), more details and color nuances
      appear, but the segmentation can also become noisy or fragmented.
    </li>
  </ul>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>1.g) 3D point cloud segmentation (airport data)</h2>
  <p>
    In this section, the 3D point cloud of an airport is segmented using k-means
    clustering. The array <code>points</code> is assumed to contain the 3D
    coordinates (x, y, z) of the point cloud.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>from mpl_toolkits.mplot3d import Axes3D  # noqa: F401

# Example with k = 4 clusters
kmeans_airport = KMeans(n_clusters=4, random_state=0).fit(points)
labels_airport = kmeans_airport.labels_

fig = plt.figure(figsize=(6, 6))
ax = fig.add_subplot(111, projection="3d")
ax.scatter(points[:, 0], points[:, 1], points[:, 2], c=labels_airport, s=2)
ax.set_title("3D point cloud segmentation (k = 4)")
plt.show()</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>1.h) Changing k for 3D point cloud segmentation</h2>
  <p>
    Here I vary the number of clusters <em>k</em> to see how the segmentation of
    the airport point cloud changes when using different values of <em>k</em>.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>for k in [3, 5, 7]:
    kmeans_air = KMeans(n_clusters=k, random_state=0).fit(points)
    labels_air = kmeans_air.labels_

    fig = plt.figure(figsize=(6, 6))
    ax = fig.add_subplot(111, projection="3d")
    ax.scatter(points[:, 0], points[:, 1], points[:, 2], c=labels_air, s=2)
    ax.set_title(f"3D point cloud segmentation (k = {k})")
    plt.show()</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p><strong>What I observe:</strong></p>
  <ul>
    <li>
      For small values of <em>k</em> (e.g. 3), the segmentation is coarse and
      large parts of the airport are grouped into only a few clusters.
    </li>
    <li>
      As <em>k</em> increases (e.g. 5 or 7), smaller structures and different
      functional areas of the airport become separated into distinct clusters.
    </li>
    <li>
      If <em>k</em> is too large, it can lead to over-segmentation where points
      that belong to the same physical structure are split into several
      clusters.
    </li>
  </ul>
</div>

</body>
</html>
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>k-means_clustering_task02</title>
  <style>
    body {
      font-family: system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif;
      margin: 20px;
      background: #f6f7fb;
      color: #222;
      line-height: 1.6;
    }
    h1, h2, h3 {
      color: #1b3a57;
    }
    .cell {
      background: #ffffff;
      border-radius: 6px;
      padding: 15px 18px;
      margin-bottom: 16px;
      box-shadow: 0 2px 4px rgba(15, 23, 42, 0.08);
    }
    .cell-title {
      font-size: 0.8rem;
      text-transform: uppercase;
      letter-spacing: 0.06em;
      color: #667085;
      margin-bottom: 6px;
    }
    .markdown {
      border-left: 4px solid #2563eb33;
    }
    .code {
      border-left: 4px solid #16a34a33;
    }
    pre {
      background: #0f172a;
      color: #e5e7eb;
      padding: 12px 14px;
      border-radius: 4px;
      overflow-x: auto;
      font-size: 0.9rem;
    }
    code {
      font-family: "Fira Code", ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
    }
    .note {
      background: #e0f2fe;
      border-left: 4px solid #0ea5e9;
      padding: 10px 12px;
      border-radius: 4px;
      font-size: 0.9rem;
      margin-top: 6px;
    }
  </style>
</head>
<body>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h1>Week 10 – Task 2: K-means clustering on apartments data</h1>
  <p>
    This HTML document contains my work for <strong>Task 2</strong> of Week 10
    (k-means clustering on the apartments data set, using an extended feature
    space, the elbow method, model attributes and the Silhouette Score).
  </p>
  <p>
    The code cells are meant to be used inside the original section
    <em>“4.) Finding clusters in the apartments data …”</em> of the provided
    notebook. I extended the example according to sub-tasks (b)–(g).
  </p>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>2.b) Extend the data frame X3 with additional numerical variables</h2>
  <p>
    In the prepared example only the variables <code>rooms</code>,
    <code>area</code> and <code>price_per_m2</code> were used. Here I extend
    the data frame <code>X3</code> by including additional numerical variables
    such as <code>lat</code>, <code>lon</code>, <code>pop_dens</code> and
    <code>tax_income</code>, assuming that these columns are available in the
    <code>apartments_df</code> data frame.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code># apartments_df is assumed to be the full apartments DataFrame loaded earlier

X3 = apartments_df[[
    "rooms",
    "area",
    "price_per_m2",
    "lat",
    "lon",
    "pop_dens",
    "tax_income"
]].dropna()

X3.head()</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>2.c) Elbow method for the extended data frame X3</h2>
  <p>
    I apply the elbow method on the extended data frame <code>X3</code> to find
    a suitable number of clusters <em>k</em>. The inertia (sum of squared
    distances to the assigned cluster centers) is plotted for different values
    of <em>k</em>.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

inertia = []
K = range(2, 12)

for k in K:
    kmeans = KMeans(n_clusters=k, random_state=0).fit(X3)
    inertia.append(kmeans.inertia_)

plt.figure()
plt.plot(K, inertia, marker="o")
plt.title("Elbow Method - Extended Apartments Data")
plt.xlabel("k")
plt.ylabel("Inertia")
plt.show()</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p><strong>Interpretation:</strong></p>
  <ul>
    <li>
      I examine the plot and look for the “elbow”, where the decrease in
      inertia slows down noticeably.
    </li>
    <li>
      This value is my candidate for the optimal number of clusters
      <em>k</em>. In my case the elbow appears around
      <strong>k = 4–5</strong>.
    </li>
  </ul>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>2.d) K-means clustering using the optimal k</h2>
  <p>
    Based on the elbow method, I choose an optimal number of clusters
    <code>optimal_k</code>. In the example below I use
    <code>optimal_k = 5</code>, but in practice this should be adapted to the
    actual elbow observed in the plot.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>optimal_k = 5  # replace with the value suggested by the elbow method

kmeans_apmts = KMeans(n_clusters=optimal_k, random_state=0).fit(X3)</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>2.e) Attributes of the k-means model kmeans_apmts</h2>
  <p>
    I now print the requested attributes of the model:
    <code>labels_</code>, <code>inertia_</code>,
    <code>cluster_centers_</code> and <code>feature_names_in_</code>.
    These are documented in <code>sklearn.cluster.KMeans</code> under
    <em>Attributes</em>.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>print(kmeans_apmts.labels_, "\n")
print(kmeans_apmts.inertia_, "\n")
print(kmeans_apmts.cluster_centers_, "\n")
print(kmeans_apmts.feature_names_in_)</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p><strong>Explanation of the attributes:</strong></p>
  <ul>
    <li>
      <strong><code>labels_</code></strong>:
      An array with one entry per apartment, containing the cluster index
      (0, 1, …, k–1) to which each apartment was assigned.
    </li>
    <li>
      <strong><code>inertia_</code></strong>:
      The final value of the k-means objective function, i.e. the sum of
      squared distances of each point to its assigned cluster center. Lower
      values indicate more compact clusters.
    </li>
    <li>
      <strong><code>cluster_centers_</code></strong>:
      A matrix of shape <code>(k, n_features)</code> containing the coordinates
      of each cluster centroid in the feature space
      (<code>rooms</code>, <code>area</code>, <code>price_per_m2</code>,
      <code>lat</code>, <code>lon</code>, <code>pop_dens</code>,
      <code>tax_income</code>).
    </li>
    <li>
      <strong><code>feature_names_in_</code></strong>:
      The names of the input features that were used to fit the model; here
      it lists the columns that form <code>X3</code>.
    </li>
  </ul>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <h2>2.g) Silhouette Score for kmeans_apmts</h2>
  <p>
    I compute the <strong>Silhouette Score</strong> for the chosen value of
    <em>k</em>. A higher score (closer to 1) indicates better-defined and better
    separated clusters.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code>from sklearn.metrics import silhouette_score

sil_score = silhouette_score(X3, kmeans_apmts.labels_)
print("Silhouette Score for k =", optimal_k, ":", sil_score)</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p>
    To check the consistency between the elbow method and the Silhouette Score,
    I can also calculate the Silhouette Score for several values of
    <em>k</em> and compare them.
  </p>
</div>

<div class="cell code">
  <div class="cell-title">Code cell</div>
  <pre><code># Optional: Silhouette analysis for several k values
silhouette_scores = {}

for k in range(2, 12):
    kmeans_tmp = KMeans(n_clusters=k, random_state=0).fit(X3)
    score = silhouette_score(X3, kmeans_tmp.labels_)
    silhouette_scores[k] = score
    print(f"k = {k}, Silhouette Score = {score}")</code></pre>
</div>

<div class="cell markdown">
  <div class="cell-title">Markdown cell</div>
  <p><strong>Comment:</strong></p>
  <ul>
    <li>
      I compare the scores printed above with the chosen
      <code>optimal_k</code> from the elbow method.
    </li>
    <li>
      If the Silhouette Score is highest (or close to the highest) at
      <code>optimal_k</code>, then the elbow method and Silhouette analysis are
      consistent.
    </li>
    <li>
      If another value of <em>k</em> clearly has a higher Silhouette Score, I
      could mention that the Silhouette analysis suggests a slightly different
      number of clusters than the elbow method.
    </li>
  </ul>
</div>

</body>
</html>
